File: src/main/java/com/pintu/coroutines/Webservice.java
package com.pintu.coroutines;

import com.google.gson.Gson;
import org.jetbrains.annotations.NotNull;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.time.Duration;
import java.time.Instant;
import java.util.Arrays;
import java.util.List;
import java.util.Map;


class Exchange {
    private Map<String, Double> rates;
   private String base;
   private String date;

    public Map<String, Double> getRates() {
        return rates;
    }

    public void setRates(Map<String, Double> rates) {
        this.rates = rates;
    }

    public String getBase() {
        return base;
    }

    public void setBase(String base) {
        this.base = base;
    }

    public String getDate() {
        return date;
    }

    public void setDate(String date) {
        this.date = date;
    }
}

public class Webservice {
    private static final String BASE = "USD";
    public static void main(String[] args) {
        try {
            final Instant start = Instant.now() ;
            List<String> symbols = Arrays.asList("INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY","INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY","INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
                    "SGD", "JPY", "INR", "CAD", "SGD", "JPY");
            symbols.parallelStream().forEach(Webservice::printExchangeRates);
            System.out.println(Duration.between(start, Instant.now()).toMillis());
            Thread.sleep(500);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private static void printExchangeRates(String symbol) {
        final String data;
        try {
            data = getExchangeRates(symbol);
            Gson gson = new Gson();
            Exchange exchange = gson.fromJson(data, Exchange.class);
            System.out.println(symbol +"  " + exchange.getRates().get(symbol));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @NotNull
    private static String getExchangeRates(final String symbol) throws IOException {
        final URL url = new URL("https://api.exchangeratesapi.io/latest?base=" + BASE + "&symbols=" + symbol);
        final BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()));
        return reader.lines().findFirst().get();
    }
}

File: src/main/kotlin/com/pintu/coroutines/WebService.kt
package com.pintu.coroutines

import com.google.gson.Gson
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.time.Duration
import java.time.Instant


data class ExchangeKt(
    var rates: Map<String, Double>,
    var base: String,
    var date: String
)


val BASE = "USD"

fun main() {
    try {
        val start = Instant.now()
        val symbols = getSymbols()
        val deferredRates = symbols.map { GlobalScope.async { getExchangeRates(it) } }
        GlobalScope.launch {
            deferredRates.forEach { println(it.await()) }
        }
        println(Duration.between(start, Instant.now()).toMillis())
        Thread.sleep(5000)
    } catch (ex: Exception) {
        ex.printStackTrace()
    }
}

fun getSymbols(): Sequence<String> {
    return sequenceOf(
        "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY","INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY","INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD", "SGD", "JPY", "INR", "CAD",
        "SGD", "JPY", "INR", "CAD", "SGD", "JPY"
    )
}


fun getExchangeRates(symbol: String): Double? {
    val data: String
    return try {
        data = getExchangeRate(symbol)
        val gson = Gson()
        gson.fromJson(data, ExchangeKt::class.java).rates[symbol]
    } catch (e: IOException) {
        e.printStackTrace()
        0.0
    }

}

@Throws(IOException::class)
fun getExchangeRate(symbol: String): String {
    val url = URL("https://api.exchangeratesapi.io/latest?base=$BASE&symbols=$symbol")
    val reader = BufferedReader(InputStreamReader(url.openStream()))
    return reader.lines().findFirst().get()
}

